File: core/js/src/utils/topological_sort.ts
// Copyright (c) 2020 The DashQL Authors

/// A native heap
export class TopologicalSort {
    /// The entries
    _entries: Uint32Array;
    /// The index
    _index: Uint32Array;
    /// The size
    _size: number;

    public constructor(entries: [number, number][]) {
        this._entries = new Uint32Array(2 * entries.length);
        this._index = new Uint32Array(entries.length);
        this._size = entries.length;
        for (let i = 0; i < entries.length; ++i) {
            this._entries[2 * i] = entries[i][0];
            this._entries[2 * i + 1] = entries[i][1];
            this._index[entries[i][0]] = i;
        }
    }

    /// Access element
    protected value(index: number) {
        return this._entries[2 * index];
    }
    /// Access rank
    protected rank(index: number) {
        return this._entries[2 * index + 1];
    }
    /// Swap two positions
    protected swapAt(i: number, j: number) {
        this._index[this.value(i)] = j;
        this._index[this.value(j)] = i;
        const v = this.value(i);
        const r = this.rank(i);
        this._entries[2 * i] = this._entries[2 * j];
        this._entries[2 * i + 1] = this._entries[2 * j + 1];
        this._entries[2 * j] = v;
        this._entries[2 * j + 1] = r;
    }
    /// Sift an element up
    protected siftUp(i: number) {
        for (let p = Math.floor((i - 1) / 2); i > 0 && this.rank(p) > this.rank(i);) {
            console.log(p);
            this.swapAt(i, p);
            i = p;
            p = Math.floor((i - 1) / 2);
        }
    }
    /// Sift an element down
    protected siftDown(i: number) {
        while (true) {
            const l = 2 * i + 1;
            const r = 2 * i + 2;
            const prev = i;
            if (l < this._size && this.rank(l) < this.rank(i)) {
                this.swapAt(l, i);
                i = l;
            }
            if (r < this._size && this.rank(r) < this.rank(i)) {
                this.swapAt(r, i);
                i = r;
            }
            if (prev == i) {
                break;
            }
        }
    }
    /// Is empty?
    public empty(): boolean { return this._size <= 0; }
    /// Get the min element
    public top(): number { return this._entries[0]; }
    /// Get the min element rank
    public topRank(): number { return this._entries[1]; }
    /// Pop the back of the vector
    protected popBack() { --this._size; }
    /// Pop the min element
    public pop() {
        this.swapAt(0, this._size - 1);
        this.popBack();
        this.siftDown(0);
    }
    /// Decrement a key
    public decrementKey(key: number, by: number) {
        const i = this._index[key];
        this._entries[2 * i + 1] -= Math.min(this._entries[2 * i + 1], by);
        this.siftUp(i);
    }
    /// Get the current rank of a key
    public findRank(key: number) {
        return this.rank(this._index[key]);
    }
}

File: libs/dashql-core/src/utils/topological_sort.rs
use std::{collections::HashMap, hash::Hash};

#[derive(Debug)]
pub struct TopologicalSort<V>
where
    V: PartialEq + Eq + Hash + Clone,
{
    entries: Vec<(V, usize)>,
    index: HashMap<V, usize>,
}

impl<V> TopologicalSort<V>
where
    V: PartialEq + Eq + Hash + Clone + std::fmt::Debug,
{
    /// Constructor
    pub fn new(mut entries: Vec<(V, usize)>) -> Self {
        entries.sort_by_key(|(_, prio)| *prio);
        let mut index = HashMap::new();
        for (i, (v, _)) in entries.iter().enumerate() {
            index.insert(v.clone(), i);
        }
        Self { entries, index }
    }

    /// Swap indices
    pub fn swap_at(&mut self, i: usize, j: usize) {
        self.index.insert(self.entries[i].0.clone(), j);
        self.index.insert(self.entries[j].0.clone(), i);
        let tmp = self.entries[i].clone();
        self.entries[i] = self.entries[j].clone();
        self.entries[j] = tmp;
    }

    /// Sift an element up
    pub fn sift_up(&mut self, mut i: usize) {
        while i != 0 && (self.entries[(i - 1) / 2].1 > self.entries[i].1) {
            let p = (i - 1) / 2;
            self.swap_at(i, p);
            i = p;
        }
    }

    /// Sift an element down
    pub fn sift_down(&mut self, mut i: usize) {
        loop {
            let l = 2 * i + 1;
            let r = 2 * i + 2;
            let prev = i;
            if l < self.entries.len() && self.entries[l].1 < self.entries[i].1 {
                i = l;
            }
            if r < self.entries.len() && self.entries[r].1 < self.entries[i].1 {
                i = r;
            }
            if prev == i {
                break;
            }
            self.swap_at(prev, i);
        }
    }
    /// Heap length
    pub fn len(&self) -> usize {
        self.entries.len()
    }
    /// Heap is empty
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }
    /// Get the min element
    pub fn top<'a>(&'a self) -> &'a (V, usize) {
        self.entries.first().unwrap()
    }
    /// Pop the min element
    pub fn pop(&mut self) {
        self.swap_at(0, self.entries.len() - 1);
        let (v, _prio) = self.entries.pop().unwrap();
        self.index.remove(&v);
        self.sift_down(0);
    }
    /// Decrement the key
    pub fn decrement_key(&mut self, k: &V) {
        if !self.index.contains_key(k) {
            return;
        }
        let i = *self.index.get(k).unwrap();
        if self.entries[i].1 > 0 {
            self.entries[i].1 -= 1;
            self.sift_up(i);
        }
    }
    /// Get the priority
    pub fn get_priority(&self, k: &V) -> usize {
        let i = self.index.get(k).copied().unwrap();
        self.entries[i].1
    }
}
